Use the m2m field for inventory source credentials

AlanCoding opened the pull request:
This removes the `credential` field on InventorySource and their jobs (inventory update).

The replacement already exists as the `credentials` field on the UnifiedJobTemplate polymorphic model that InventoryUpdate inherits from - same for UnifiedJob.

There isn't any major immediate benefit that comes from this. The reason we want this is so that we can smoothly transition into doing multi-vault. We could also in the future allow multiple credentials for inventory updates. For custom credential types, this can be done with a trivial amount of work, but I'm not diving in to claim that as a feature just yet.

This probably still needs some more testing, and I might be working on some Jenkins failure. But the UI works (I kept the existing data unchanged), and stuff I put in there pre-migration is working just fine.

EDIT: Also, still need RBAC...

EDIT: alright, this is starting to look reasonably complete, got openstack example run verified

EDIT: no wait, no I don't

AlanCoding committed:
use the m2m field for inventory source creds

AlanCoding commented:
...been trying to get the serializer fully tested and wrapped up.

Doing this, I've realized that the updated contract is actually somewhat nasty for a client to use. I've tried to just apply our existing patterns, but it's hard to fully defend them here. The problem is that an inventory source can only use exactly 1 cloud credential, and this credential is almost always required by the type of the inventory source.

So consider the process of changing the type of an inventory source from one cloud type to another.

### Old method

PATCH with

```json
{
  "source": "stackystack",
  "credential": 42
}
```

Where "stackystack" is the new type of inventory source, and 42 is a credential that corresponds to that type.

### New method

Since it's many-to-many, we'll need to start off just changing the source

PATCH

```json
{
  "source": "stackystack"
}
```

Now, we should apply the credential. As the client, I POST `{"id": 42}` to the credentials list. But when I do that, I get a message that this inventory source already has a cloud credential associated with it, which is of the previous type. That means that I would first have to remove that.

POST to `/api/v2/inventory_sources/9/credentials/`
```json
{
  "id": 34,
  "disassociate": true
}
```

Where 9 is the id of the inventory source and 34 is the id of the credential that corresponds to its _old_ type.

After that, we can associate the new one.

POST to `/api/v2/inventory_sources/9/credentials/`
```
{
  "id": 42
}
```

This makes it a somewhat frustrating 3-step process. @ryanpetrello how does this make you feel? Any suggestions on the general structure of this contract?

AlanCoding commented:
Link #328, which could be re-written to work with multi-vault after this change. Major qualifier - this will involve extremely invasive work to the inventory import command to use pexpect.

But I have no idea what to do about hiding secrets.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2183,8 +2257,6 @@ def validate_credential_type(self, credential_type):
             for rel in (
                 'ad_hoc_commands',
                 'insights_inventories',
-                'inventorysources',
```
If there's not an explicit test for this scenario, can we add one to make our lives easier when 3.2.3 merges (I want to make sure we don't regress on that merge)?

ryanpetrello commented on awx/main/migrations/_multi_cred.py:
```diff
@@ -32,3 +32,25 @@ def migrate_workflow_cred_reverse(app, schema_editor):
+    InventoryUpdate = app.get_model('main', 'InventoryUpdate')
+    InventorySource = app.get_model('main', 'InventorySource')
+
+    for cls in (InventoryUpdate, InventorySource):
```
This is dedication ;)

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -915,8 +915,11 @@ def run(self, pk, isolated_host=None, **kwargs):
+                credentials = [instance.get_cloud_credential()]
             elif hasattr(instance, 'credential'):
-                # once other UnifiedJobs (project updates, inventory updates)
+                # once other UnifiedJobs (project updates)
```
Does it make sense to just change this now to something more explicit, e.g.,

`elif isinstance(instance, ProjectUpdate)` ?

AlanCoding replied:
answering this for myself...

```python
In [6]: for cls in UnifiedJobTemplate.__subclasses__():
   ...:     try:
   ...:         cls._meta.get_field('credential')
   ...:         print str(cls) + ' has credential '
   ...:     except:
   ...:         print str(cls) + ' does not have it'
   ...:         
<class 'awx.main.models.projects.Project'> has credential 
<class 'awx.main.models.inventory.InventorySource'> does not have it
<class 'awx.main.models.jobs.JobTemplate'> does not have it
<class 'awx.main.models.jobs.SystemJobTemplate'> does not have it
<class 'awx.main.models.workflow.WorkflowJobTemplate'> does not have it
```

yeah, okay, so that really is down to _just_ Project.

I think we only use this list of `credentials` for injecting custom credentials. I don't know if there's any supported use case for projects, but that still seems sensible in a certain way. Maybe the git or hg modules could use credentials in a way we didn't anticipate.

ryanpetrello commented on awx/main/tests/functional/api/test_credential.py:
```diff
@@ -1426,9 +1426,9 @@ def test_field_removal(put, organization, admin, credentialtype_ssh, version, pa
-    ['inventorysources', InventorySource()],
     ['unifiedjobs', Job()],
     ['unifiedjobtemplates', JobTemplate()],
+    ['unifiedjobtemplates', InventorySource()],
```
Disregard my comment above - thanks for adding this.

ryanpetrello reviewed (approved):
This looks great - thanks for putting in the work on this.  It will be nice to lose this API long-term and have a unified credentials list in the places where it makes sense.

AlanCoding committed:
docs and review change for IS multivault

Mention inventory sources /credentials/ endpoint in docs
Also change means of identifying projects for the purose
of injecting custom credentials

ryanpetrello commented on docs/multi_credential_assignment.md:
```diff
@@ -220,3 +220,16 @@ POST /api/v2/job_templates/N/launch/
+
+Most cloud sources will continue to adhere to the constraint that they
+must have a single credential that corresponds to their cloud type.
+However, this relationship allows users to associate multiple vault
```
There's one important question we maybe didn't answer succinctly in our meeting.  Given:

> However, this relationship allows users to associate multiple vault credentials of different ids to inventory sources.

Are we comfortable shrugging our shoulders about what this _means_ until we actually implement it later?  If somebody actually uses this API to assign multi vault, what does it _do_?  What _should_ it do?  Are we comfortable with ambiguity?  If not, maybe we want to carve out the multi-Vault-specific stanzas in this PR until we're ready to thoroughly implement them?

AlanCoding replied:
Discussion is in #223, and you go down one of two roads.

Either you say that (1) it's okay to import vaulted content into our current system, so it goes into group / host variables plaintext, or (2) this content needs to be redacted / encrypted like survey passwords.

(1) is re-implementing the prior PR on top of this... probably a day to have a PR due to the need to switch to pexpect.

Option (2) won't be in the cards for a _long_ time. Inventory imports run through `ansible-inventory`, and no mechanism exists to know which values were from vault. So we're talking about a new feature request for Ansible core, then adding the flag to use it, developing a solution for old versions _without_ the feature. We could say that the presence of _any_ vault credentials means that all variables need to be saved encrypted, which is... yuck.

I would like to give a little push toward option (1). Firstly, consider the quite substantial community demand expressed. One comment in that discussion got _15_ 👍, and I think this kind of response should factor into our plan to at least some degree. AFAIK, nothing about the community feature request necessitates encryption. Option (1) also could latter be expanded to (2), in a future world where the pre-reqs are available and users have asked for it. If this gets into the UI (which I also realize is a discussion), then documentation needs to be self-evident end-to-end so that users know they will _see_ variables right away after import.

ryanpetrello replied:
Understood.  What I'm asking is do we want to merge _this_ PR without solutions #1 and #2 existing?  Are we comfortable with the API supporting vault assignment in an ambiguous way until there's an actual solution for #223?

AlanCoding replied:
I hear you there. For this PR, I could prohibit vault creds and put in a TODO about that so we don't have any transient inconsistent behavior. That would help to keep these different tasks separate. This still feels right to me as 2 separate PRs.

I've been hitting up @jakemcdermott and @jladdjr about the possible feature drop.

ryanpetrello replied:
Yea, I agree that the actual implementation for Vault makes sense in a separate PR - this PR is good as is, but perhaps we could block Vault for now to avoid confusion.

AlanCoding replied:
Right, this'll get dropped back in as part of that.

```python
# Vault credentials are only exclusive with others of same ID
if sub.unique_hash() in [cred.unique_hash() for cred in parent.credentials.all()]:
    return {"msg": _(
        "A credential of type {credential_type} is already assigned to this inventory source."
    ).format(credential_type=sub.unique_hash(display=True))}
```

AlanCoding commented:
Totally missed the connection to https://github.com/ansible/awx/issues/277
